File: ts/packages/marginfi-client/src/bank.ts
import { PublicKey } from "@solana/web3.js";
import BigNumber from "bignumber.js";
import { PARTIAL_LIQUIDATION_FACTOR } from "./constants";
import { BankData, LendingSide, MarginRequirementType } from "./types";
import { wrappedI80F48toBigNumber } from "./utils/helpers";

/**
 * Bank struct mirroring on-chain data
 * Contains the state of the marginfi group.
 */
class Bank {
  public readonly scalingFactorC: BigNumber;
  public readonly fixedFee: BigNumber;
  public readonly interestFee: BigNumber;
  public readonly depositAccumulator: BigNumber;
  public readonly borrowAccumulator: BigNumber;
  public readonly lastUpdate: Date;
  public readonly totalDepositsRecord: BigNumber;
  public readonly totalBorrowsRecord: BigNumber;
  public readonly mint: PublicKey;
  public readonly vault: PublicKey;
  public readonly bankAuthorityBump: number;
  public readonly insuranceVault: PublicKey;
  public readonly insuranceVaultAuthorityBump: number;
  public readonly feeVault: PublicKey;
  public readonly feeVaultAuthorityBump: number;
  public readonly initMarginRatio: BigNumber;
  public readonly maintMarginRatio: BigNumber;

  constructor(data: BankData) {
    this.scalingFactorC = wrappedI80F48toBigNumber(data.scalingFactorC, 0);
    this.fixedFee = wrappedI80F48toBigNumber(data.fixedFee, 0);
    this.interestFee = wrappedI80F48toBigNumber(data.interestFee, 0);
    this.depositAccumulator = wrappedI80F48toBigNumber(data.depositAccumulator, 0);
    this.borrowAccumulator = wrappedI80F48toBigNumber(data.borrowAccumulator, 0);
    this.lastUpdate = new Date(data.lastUpdate.toNumber());
    this.totalDepositsRecord = wrappedI80F48toBigNumber(data.totalDepositsRecord, 0);
    this.totalBorrowsRecord = wrappedI80F48toBigNumber(data.totalBorrowsRecord, 0);
    this.mint = data.mint;
    this.vault = data.vault;
    this.bankAuthorityBump = data.bankAuthorityBump;
    this.insuranceVault = data.insuranceVault;
    this.insuranceVaultAuthorityBump = data.insuranceVaultAuthorityBump;
    this.feeVault = data.feeVault;
    this.feeVaultAuthorityBump = data.feeVaultAuthorityBump;
    this.initMarginRatio = wrappedI80F48toBigNumber(data.initMarginRatio, 0);
    this.maintMarginRatio = wrappedI80F48toBigNumber(data.maintMarginRatio, 0);
  }

  get nativeDepositBalance(): BigNumber {
    return this.computeNativeAmount(this.totalDepositsRecord, LendingSide.Deposit).shiftedBy(-6);
  }

  get nativeBorrowBalance(): BigNumber {
    return this.computeNativeAmount(this.totalBorrowsRecord, LendingSide.Borrow).shiftedBy(-6);
  }

  public computeNativeAmount(record: BigNumber, side: LendingSide): BigNumber {
    if (side === LendingSide.Borrow) {
      return record.times(this.borrowAccumulator);
    } else if (side === LendingSide.Deposit) {
      return record.times(this.depositAccumulator);
    } else {
      throw Error(`Unknown lending side: ${side}`);
    }
  }

  public computeRecordAmount(record: BigNumber, side: LendingSide): BigNumber {
    if (side === LendingSide.Borrow) {
      return record.div(this.borrowAccumulator);
    } else if (side === LendingSide.Deposit) {
      return record.div(this.depositAccumulator);
    } else {
      throw Error(`Unknown lending side: ${side}`);
    }
  }

  public marginRatio(type: MarginRequirementType): BigNumber {
    switch (type) {
      case MarginRequirementType.Init:
        return this.initMarginRatio;
      case MarginRequirementType.Maint:
        return this.maintMarginRatio;
      case MarginRequirementType.PartialLiquidation:
        return PARTIAL_LIQUIDATION_FACTOR.times(this.initMarginRatio.minus(this.maintMarginRatio)).plus(
          this.maintMarginRatio
        );
      default:
        throw Error(`Unknown margin requirement type: ${type}`);
    }
  }
}

export default Bank;

File: python/marginpy/python/marginpy/bank.py
from datetime import datetime

from marginpy.constants import COLLATERAL_SCALING_FACTOR, PARTIAL_LIQUIDATION_FACTOR
from marginpy.types import BankData, LendingSide, MarginRequirement
from marginpy.utils.data_conversion import wrapped_fixed_to_float
from solana.publickey import PublicKey


class Bank:
    """
    Bank struct mirroring on-chain data.

    Contains the state of the marginfi group.
    """

    scaling_factor_c: float
    fixed_fee: float
    interest_fee: float
    deposit_accumulator: float
    borrow_accumulator: float
    last_update: datetime
    total_deposits_record: float
    total_borrows_record: float
    mint: PublicKey
    vault: PublicKey
    vault_authority_pda_bump: int
    insurance_vault: PublicKey
    insurance_vault_authority_pda_bump: int
    fee_vault: PublicKey
    fee_vault_authority_pda_bump: int
    init_margin_ratio: float
    maint_margin_ratio: float
    account_deposit_limit: float
    lp_deposit_limit: float

    def __init__(self, data: BankData) -> None:
        self.scaling_factor_c = wrapped_fixed_to_float(data.scaling_factor_c)
        self.fixed_fee = wrapped_fixed_to_float(data.fixed_fee)
        self.interest_fee = wrapped_fixed_to_float(data.interest_fee)
        self.deposit_accumulator = wrapped_fixed_to_float(data.deposit_accumulator)
        self.borrow_accumulator = wrapped_fixed_to_float(data.borrow_accumulator)
        self.last_update = datetime.fromtimestamp(data.last_update)
        self.total_deposits_record = (
            wrapped_fixed_to_float(data.total_deposits_record)
            / COLLATERAL_SCALING_FACTOR
        )
        self.total_borrows_record = (
            wrapped_fixed_to_float(data.total_borrows_record)
            / COLLATERAL_SCALING_FACTOR
        )
        self.mint = data.mint
        self.vault = data.vault
        self.vault_authority_pda_bump = data.vault_authority_pda_bump
        self.insurance_vault = data.insurance_vault
        self.insurance_vault_authority_pda_bump = (
            data.insurance_vault_authority_pda_bump
        )
        self.insurance_vault_outstanding_transfers = (
            data.insurance_vault_outstanding_transfers
        )
        self.fee_vault = data.fee_vault
        self.fee_vault_authority_pda_bump = data.fee_vault_authority_pda_bump
        self.init_margin_ratio = wrapped_fixed_to_float(data.init_margin_ratio)
        self.maint_margin_ratio = wrapped_fixed_to_float(data.maint_margin_ratio)
        self.account_deposit_limit = (
            wrapped_fixed_to_float(data.account_deposit_limit)
            / COLLATERAL_SCALING_FACTOR
        )
        self.lp_deposit_limit = (
            wrapped_fixed_to_float(data.lp_deposit_limit) / COLLATERAL_SCALING_FACTOR
        )

    def compute_native_amount(
        self,
        record_amount: float,
        side: LendingSide,
    ) -> float:
        """
        Converts the provided record amount to its native amount respective to current accumulator.

        Raises:
            Exception: unknown side
        """

        if side == LendingSide.BORROW:
            return record_amount * self.borrow_accumulator

        if side == LendingSide.DEPOSIT:
            return record_amount * self.deposit_accumulator

        raise Exception(f"Unknown lending side: {side}")

    def compute_record_amount(self, native_amount: float, side: LendingSide) -> float:
        """
        Converts the provided native amount to its record amount respective to current accumulator.

        Raises:
            Exception: unknown side
        """

        if side == LendingSide.BORROW:
            return native_amount / self.borrow_accumulator

        if side == LendingSide.DEPOSIT:
            return native_amount / self.deposit_accumulator

        raise Exception(f"Unknown lending side: {side}")

    def compute_margin_ratio(self, mreq_type: MarginRequirement) -> float:
        if mreq_type is MarginRequirement.INITIAL:
            return self.init_margin_ratio

        if mreq_type is MarginRequirement.MAINTENANCE:
            return self.maint_margin_ratio

        if mreq_type is MarginRequirement.PARTIAL_LIQUIDATION:
            return self.maint_margin_ratio + PARTIAL_LIQUIDATION_FACTOR * (self.init_margin_ratio - self.maint_margin_ratio)

        raise Exception(f"Unknown margin requirement type: {mreq_type}")
